content/en/blog/extend-falco-inputs-with-a-plugin-the-basics.md

```diff
@@ -0,0 +1,700 @@
+
+### The functions and methods
+
+#### `main()`
+
+The `main()` function is mandatory for any `go` program, but because we'll build the `plugin` as a library for the *Falco plugin framework* which is written in `C`, we can let it empty.
+
+```go
+// main is mandatory but empty, because the plugin will be used as C library by Falco plugin framework
+func main() {}
+```
+
+#### `init()`
+
+The `init()` function is used for registering our plugin to the `Falco plugin framework`, as a [`source`](https://falco.org/docs/plugins/#source-plugin) and [`extractor`](https://falco.org/docs/plugins/#extractor-plugin).
+
+```go
+// init function is used for referencing our plugin to the Falco plugin framework
+func init() {
+	p := &DockerPlugin{}
+	extractor.Register(p)
+	source.Register(p)
+}
+```
+
+#### `Info()`
+
+This method is mandatory, and all plugins must respect that. It allows the `Falco plugin framework` to have all intel about the plugin itself:
+
+```go
+// Info displays information of the plugin to Falco plugin framework
+func (dockerPlugin *DockerPlugin) Info() *plugins.Info {
+	return &plugins.Info{
+		ID:                 6,
+		Name:               "docker",
+		Description:        "Docker Events",
+		Contact:            "github.com/falcosecurity/plugins/",
+		Version:            "0.1.0",
+		RequiredAPIVersion: "0.3.0",
+		EventSource:        "docker",
+	}
+}
+```
+
+Here some details:
+* `ID`: must be unique among all plugins, it's used by the framework in captures to know which `plugin` is the `source` of events. It's also important for avoiding collisions if you want to share your plugin in the [registry](https://github.com/falcosecurity/plugins). See [documentation](https://falco.org/docs/plugins/#plugin-event-ids) for more details.
+* `Name`: the name of our plugin, will be used in `plugins` section of  `falco.yaml`
+* `EventSource`: this represents the value we'll set in `Falco` rules for mapping, in our case, all rules we'll set will have `source: docker`
+
+#### `Init()`
+
+This method (:warning: different from the function `init()`) will be the first one called by the *Falco plugin framework*, we use it for setting default values for `DockerPlugin` attributes. In our case, these default values are overridden by the value of `init_config:` from `falco.yaml` config file, see []().
+ 
+```go
+// Init is called by the Falco plugin framework as first entry
+// we use it for setting default configuration values and mapping
+// values from `init_config` (json format for this plugin)
+func (dockerPlugin *DockerPlugin) Init(config string) error {
+	dockerPlugin.FlushInterval = 2
+	json.Unmarshal([]byte(config), &dockerPlugin)
```

Comment:
I would add a little comment here to specify why we don't check the unmarshalling error. In this case, we don't need to check it because the configuration is only composed by one field and we don't care if a wrong config is passed, or if the field is not specified, because the default will silently be used. I would not want this error-ignoring approach with the `InitSchema()` feature, which instead ensures that errors are checked right by the framework.

Reply:
I think I will expose `InitSchema()` in the advanced post, wdyt?

Reply:
Yeah definitely! But still ignoring the error is not a very good practice here, because in this case that would lead to the user using the default `FlushInterval` value without being noticed that they submitted a wrongly-formatted configuration. So I think one improvement here can be either:
- We catch the error but not return it (this case is so simple that I don't see a point in breaking the init for this), and log it as a warning
- Not catching the error (like we do now), but leave a little comment stating that this choice is for simplicity

I think it is important to give readers a good example, and error handling is something to put some attention on IMO! Just a thought!

Reply:
Ok for adding the error handling, is there a way to pass to the framework an error for making falco printing it? (to keep same format)

Reply:
`Init()` can return an error for this exact purpose. Now, returning an error at plugin initialization will stop the loading process, and Falco would stop too, which is intended behavior. If you don't want Falco to stop, then you should just log the error in `Init()` without returning it.

Reply:
```
Runtime error: cannot load plugin /usr/share/falco/plugins/libdocker.so: Could not initialize plugin: json: cannot unmarshal string into Go struct field DockerPlugin.flushInterval of type uint64. Exiting.
```

done